```rust
error[E0596]: cannot borrow immutable local variable `entry` as mutable
   --> components/style/custom_properties.rs:117:17
    |
116 |             Entry::Occupied(entry) => {
    |                             ----- consider changing this to `mut entry`
117 |                 entry.insert(value);
    |                 ^^^^^ cannot borrow mutably

error: aborting due to previous error

error: Could not compile `style`.
```

--- a/components/style/custom_properties.rs
+++ b/components/style/custom_properties.rs
@@ -678,7 +678,7 @@ fn substitute_all(custom_properties_map: &mut CustomPropertiesMap) {
         // Some shortcut checks.
         let (name, value) = if let Some(value) = context.map.get(&name) {
             // This variable has been resolved. Return the signal value.
-            if value.references.is_empty()  || context.invalid.contains(&name) {
+            if value.references.is_empty() || context.invalid.contains(&name) {
                 return None;
             }
             // Whether this variable has been visited in this traversal.

--- a/components/style/custom_properties.rs
+++ b/components/style/custom_properties.rs
@@ -259,9 +259,7 @@ impl VariableValue {
         debug_assert!(variable.references.is_empty());
         self.push(&variable.css, variable.first_token_type, variable.last_token_type)
     }
-}
 
-impl VariableValue {
     /// Parse a custom property value.
     pub fn parse<'i, 't>(
         input: &mut Parser<'i, 't>,

--- a/components/style/custom_properties.rs
+++ b/components/style/custom_properties.rs
@@ -8,6 +8,7 @@
 
 use Atom;
 use cssparser::{Delimiter, Parser, ParserInput, SourcePosition, Token, TokenSerializationType};
+use hash::map::Entry;
 use precomputed_hash::PrecomputedHash;
 use properties::{CSSWideKeyword, DeclaredValue};
 use selector_map::{PrecomputedHashSet, PrecomputedHashMap};
@@ -105,11 +106,21 @@ where
     }
 
     /// Insert a new key-value pair.
+    ///
+    /// TODO(emilio): Remove unused_mut when Gecko and Servo agree in whether
+    /// it's necessary.
+    #[allow(unused_mut)]
     pub fn insert(&mut self, key: K, value: V) {
-        if !self.values.contains_key(&key) {
-            self.index.push(key.clone());
+        let OrderedMap { ref mut index, ref mut values } = *self;
+        match values.entry(key) {
+            Entry::Vacant(mut entry) => {
+                index.push(entry.key().clone());
+                entry.insert(value);
+            }
+            Entry::Occupied(mut entry) => {
+                entry.insert(value);
+            }
         }
-        self.values.insert(key, value);
     }
 
     /// Get a value given its key.
@@ -666,8 +677,6 @@ fn substitute_all(custom_properties_map: &mut CustomPropertiesMap) {
     ///   been completely resolved.
     /// * There is no such variable at all.
     fn traverse<'a>(name: Name, context: &mut Context<'a>) -> Option<usize> {
-        use hash::map::Entry;
-
         // Some shortcut checks.
         let (name, value) = if let Some(value) = context.map.get(&name) {
             // This variable has been resolved. Return the signal value.

